server/jobs/job.go

```diff
@@ -0,0 +1,75 @@
+package jobs
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/lieut-data/mattermost-plugin-api/cluster"
+	"github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar"
+)
+
+// start waits until the job's initial run time, then runs the job every interval.
+// A cluster mutex is used to synchronize with other nodes that are running the same plugin.
+func (job *activeJob) start(getEnv func() mscalendar.Env, papi cluster.MutexPluginAPI) {
+	go func() {
+		defer close(job.cancelled)
+
+		// Get mutex to gain access to sensitive section
+		lock := cluster.NewMutex(papi, job.id)
+		lock.Lock()
+		defer lock.Unlock()
+
+		// Get initial run's time
+		firstRun := job.timeUntilFirstRun()
+		timer := time.NewTimer(firstRun)
+		getEnv().Logger.Debugf("%s job will run at at `%s`, in `%s`, then every `%s`.", job.id, time.Now().Add(firstRun).Format("Jan 02, 3:04PM MST"), timeDurationDisplay(firstRun), timeDurationDisplay(job.interval))
+
+		select {
+		case <-timer.C:
+			// time for the initial run
+		case <-job.cancel:
+			return
+		}
+
+		// Prepare ticker for recurring runs
+		ticker := time.NewTicker(job.interval)
+		defer func() {
+			ticker.Stop()
+		}()
+
+		// Do initial run
+		go job.work(getEnv())
+
+		// Loop through ticker
+		for {
+			select {
+			case <-ticker.C:
+				go job.work(getEnv())
```

Comment:
Seems like the best approach is to call without `go`. I was attempting to make it so if a job lasted longer than `INTERVAL`, then the next job would not be offset from its intended run time. But I think the solution was incorrect.

Maybe if a given iteration goes overtime, we just skip the next run? It would depend on the use case of the job of course.

Reply:
> Maybe if a given iteration goes overtime, we just skip the next run? 

+1